interceptor/zipkin/trace_interceptor.go

```diff
@@ -0,0 +1,399 @@
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strings"
+
+	"go.opencensus.io/trace"
+
+	zipkinmodel "github.com/openzipkin/zipkin-go/model"
+
+	commonpb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	agenttracepb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/trace/v1"
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/census-instrumentation/opencensus-service/interceptor"
+	"github.com/census-instrumentation/opencensus-service/internal"
+	"github.com/census-instrumentation/opencensus-service/spanreceiver"
+)
+
+type ZipkinInterceptor struct {
+	spanSink spanreceiver.SpanReceiver
+}
+
+var _ interceptor.TraceInterceptor = (*ZipkinInterceptor)(nil)
+var _ http.Handler = (*ZipkinInterceptor)(nil)
+
+func New(sr spanreceiver.SpanReceiver) (*ZipkinInterceptor, error) {
+	return &ZipkinInterceptor{spanSink: sr}, nil
+}
+
+func (zi *ZipkinInterceptor) StartTraceInterception(ctx context.Context, spanSink spanreceiver.SpanReceiver) error {
+	zi.spanSink = spanSink
+	return nil
+}
+
+func (zi *ZipkinInterceptor) parseAndConvertToTraceSpans(jsonBlob []byte) (reqs []*agenttracepb.ExportTraceServiceRequest, err error) {
+	var zipkinSpans []*zipkinmodel.SpanModel
+	if err = json.Unmarshal(jsonBlob, &zipkinSpans); err != nil {
+		return nil, err
+	}
+
+	// *commonpb.Node instances have unique addresses hence
+	// for grouping within a map, we'll use the .String() value
+	byNodeGrouping := make(map[string][]*tracepb.Span)
+	nodes := make([]*commonpb.Node, 0, len(zipkinSpans))
+	// Now translate them into tracepb.Span
+	for _, zspan := range zipkinSpans {
+		span, node, err := zipkinSpanToTraceSpan(zspan)
+		// TODO:(@odeke-em) record errors
+		if err == nil && span != nil {
+			key := node.String()
+			if _, firstTime := byNodeGrouping[key]; firstTime {
+				nodes = append(nodes, node)
+			}
+			byNodeGrouping[key] = append(byNodeGrouping[key], span)
+		}
+	}
+
+	// TODO: @odeke-em perhaps maintain the relative order that the spans were received in
+	for _, node := range nodes {
+		key := node.String()
+		spans := byNodeGrouping[key]
+		if len(spans) == 0 {
+			// Should never happen but nonetheless be cautious
+			// not to send blank spans.
+			continue
+		}
+		reqs = append(reqs, &agenttracepb.ExportTraceServiceRequest{
+			Node:  node,
+			Spans: spans,
+		})
+		delete(byNodeGrouping, key)
+	}
+
+	return reqs, nil
+}
+
+func (zi *ZipkinInterceptor) StopTraceInterception(ctx context.Context) error {
+	return nil
+}
+
+// processBodyIfNecessary checks the "Content-Encoding" HTTP header and if
+// a compression such as "gzip", "deflate", "zlib", is found, the body will
+// be uncompressed accordingly or return the body untouched if otherwise.
+// Clients such as Zipkin-Java do this behavior e.g.
+//    send "Content-Encoding":"gzip" of the JSON content.
+func processBodyIfNecessary(req *http.Request) io.Reader {
```

Comment:
nit: naming, this is about selecting the io.Reader according to encoding not actually processing the body

Reply:
@pjanotti thanks, it actually will gunzip or zlib deflate the body so it actually processes the body if necessary.
<img width="467" alt="screen shot 2018-10-25 at 11 08 08 am" src="https://user-images.githubusercontent.com/4898263/47520877-4747d380-d846-11e8-8230-1692f956c4af.png">

Reply:
Ah, I was incorrectly assuming that they were completely lazy...

Reply:
Great suggestion though to be conservative with the naming :)